### PR TITLE
Advertize compression algorithms support in `init` (features 32/33)

### LINKED ISSUE #811: Encodings in gossip_queries
Currently the spec requires an implementation to support both raw and zlib deflate encodings for `short_channel_ids` in `reply_channel_range` and `query_short_channel_ids` messages.

There is discrepancy of the default encoding used when replying to a query with `reply_channel_range` message. It would be nice to have clarification surrounding when zlib deflate or raw encodings should be sent in reply to a query.

A bigger request is that rust-lightning would like to not add a zlib dependency to the project. There is currently no way to signal support or lack thereof for zlib. It would be nice if a node could signal this support in Init. There is likely overlap with #804 depending on how reply sequencing is structured, so I wanted to raise this issue for discussions while that is being discussed.

### PR DESCRIPTION
**Pull request by t-bast**
Add a list of supported compression algorithms to the `init` message.
These compression algorithms may be used in several places in the future, but currently only in extended gossip queries.

Without this explicit field, it's impossible to opt-out of zlib compression.
If your node implementation does not support zlib, it will only be discovered when your peer sends you a `query_short_channel_ids` that asks for a compressed response. Your only choices are to close the connection or send back uncompressed data, which may or may not be accepted by your peer.

Fixes #811

### CODE REVIEW

File: 09-features.md [hunk 1]
```diff
@@ -26,20 +26,20 @@ The Context column decodes as follows:
...
+| 14/15 | `payment_secret`                 | Node supports `payment_secret` field                      | IN9      | `var_onion_optin`         | [Routing Onion Specification][bolt04]                   |
+| 16/17 | `basic_mpp`                      | Node can receive basic multi-part payments                | IN9      | `payment_secret`          | [BOLT #4][bolt04-mpp]                                   |
+| 18/19 | `option_support_large_channel`   | Can create large channels                                 | IN       |                           | [BOLT #2](02-peer-protocol.md#the-open_channel-message) |
+| 20/21 | `option_anchor_outputs`          | Anchor outputs                                            | IN       | `option_static_remotekey` | [BOLT #3](03-transactions.md)                           |
+| 24/25 | `zlib_compress`                  | Support for zlib compression                              | IN       | `gossip_queries_ex`       | [BOLT #7][bolt07-query]                                 |
```
Comment (ariard) [inline 09-features.md]: How do we interpret mandatory/optional bits ? If you setup the optional bit, does it signal as "I accept compressed queries" and implicitly signal raw format acceptance but doesn't enforce a preference on scids query sender ?

But if you're an upgraded node which doesn't support zlib but still wants to signal raw format acceptance how do you do to dissociate yourself from non-upgraded nodes ?

**Review by ariard — COMMENTED** [review]
Thanks for working on this.

File: 09-features.md (hunk 1 above)
Comment (t-bast) [reply 09-features.md:42]: Good point.

I'd say that optional support means I can do zlib and will comply with what senders ask for (and may ask you for zlib if you support it as well).

Mandatory support means I'll only do zlib and never uncompressed data. Is that too hardcore? I think it makes sense to have this hardcore mode, and it's up to node operators to decide whether they want to go hardcore or not (implementations will very likely only default to optional support or no support).

Comment (ariard) [reply 09-features.md]: I really agree it's up to node operators to decide what their gossip queries format, let's be conservative with your bandwidth.

But still, after this PR, how a queried node should signal "I-don't-want-to-support-zlib-but-only-raw-format" to dissociate themselves from non-upgraded nodes from a querying node viewpoint ?

Maybe we should discuss reversing the feature-bit to `option_raw_format` at next meeting, it sounds to solve the issue IMO.

Comment (TheBlueMatt) [reply 09-features.md:42]: I'd suggest using a `format_raw` bit instead - if a user is concerned about bandwidth, there should be a global bandwidth limiter provided which limits bandwidth of pings, syncs, and everything else. Not just enforcing that syncs only happen via one mechanism, which only indirectly limits bandwidth and will fall over if you get an old peer sending lots of pings or a sudden influx of peers.

Comment (t-bast) [reply 09-features.md:42]: I understand the reasoning behind reversing this to an `option_raw_format` / `format_raw`, but I find it weird to have feature bits that actually announce absence of support for something. Feature bits should announce support for additional capabilities, not reduced capabilities.

Granted, we're in this situation because we chose at the time to make zlib support mandatory, but I'm not sure we should go in that direction. I'd like to hear @rustyrussell's opinion on this, he and @sstone did the initial zlib/gossip_queries work.

**Comment by t-bast** [comment]
I've thought about this more, and I'm not a fan of any kind of `option_raw_format`: feature bits should not be used to advertise a lack of functionality. But I'm not a fan of an explicit `zlib_compress` feature bit either.

Another angle that could be explored for this would be a new `tlv` field in the `init` message:

```
1. type: 3 (`compression_methods`)
2. data:
        * [`tu64`:`supported_algorithms`]
```

Each bit indicates support for a specific compression algorithm.
The only supported values for now would be `0x00` (only uncompressed data) and `0x01` (support for zlib compression).
I don't expect us to ever support a lot of compression options, but since we have truncated uint64 available in all implementations, this gives us 64 potential algorithms that can be used (more than enough).

Nodes then need to remember what compression options are supported by their peers and act accordingly when doing gossip queries. If our peer set `0x00`, we should only use the raw format. If he set `0x01` we can use either raw or zlib format.

What do you think @TheBlueMatt @ariard @bmancini55? Should I amend this PR to go in that direction?

**Comment by bmancini55** [comment]
Thanks @t-bast!

I agree, `option_raw_format` seems clunky. I do think `zlib_compress` could work if we follow similar rules that you just outlined, basically defaulting to using `raw` in the absence of feature agreement.

That said, I like your new TLV proposal. It makes sense that supported compression would be indicated in the same way the encoding flags are used.

One suggestion would be to put the TLV and associated logic under an `option_compression` feature. This way it would be clear which nodes are legacy and which support these cascading compression rules.  Peer selection could then make determinations beforehand to connect to nodes that support these rules.

**Comment by t-bast** [comment]
> One suggestion would be to put the TLV and associated logic under an option_compression feature. This way it would be clear which nodes are legacy and which support these cascading compression rules. Peer selection could then make determinations beforehand to connect to nodes that support these rules.

Done in https://github.com/lightningnetwork/lightning-rfc/pull/825/commits/673c912e1a01c317c9d82f8e8b2cbe7a1f3c66c9

File: 09-features.md (hunk 1 above)
Comment (t-bast) [reply 09-features.md]: I went for a slightly different approach in https://github.com/lightningnetwork/lightning-rfc/pull/825/commits/673c912e1a01c317c9d82f8e8b2cbe7a1f3c66c9

File: 01-messaging.md [hunk 2]
```diff
@@ -258,10 +258,19 @@ The `features` field MUST be padded to bytes with 0s.
...
         * [`...*chain_hash`:`chains`]
-
+    1. type: 3 (`compression_algorithms`)
+    2. data:
+        * [`...*byte`:`supported_algorithms`]
 
 The optional `networks` indicates the chains the node is interested in.
 
+The optional `compression_algorithms` is a bitfield indicating what compression
+algorithms the node supports for [extended gossip queries](07-routing-gossip.md#query-messages).
```
Comment (bmancini55) [inline 01-messaging.md:268]: Should this be `gossip queries` instead of `extended gossip queries` since this change would impact the SCIDs in normal `reply_channel_range` and `query_short_channel_ids` messages?

Comment (t-bast) [reply 01-messaging.md:268]: Good point, I'll fix this.

**Review by TheBlueMatt — APPROVED** [review]
Looks good to me.

File: 01-messaging.md (hunk 2 above)
Comment (t-bast) [reply 01-messaging.md]: Done in https://github.com/lightningnetwork/lightning-rfc/pull/825/commits/8548f2f90e2459f38f0563c8abd2f46f926f4043

**Comment by t-bast** [comment]
Rebased and clarified the case where no compression algorithm is supported in https://github.com/lightningnetwork/lightning-rfc/pull/825/commits/c4322b74f4718bb1136eea8bf0c7f31b7191daef

File: 01-messaging.md [hunk 3]
```diff
@@ -258,10 +258,19 @@ The `features` field MUST be padded to bytes with 0s.
...
+        * [`...*byte`:`supported_algorithms`]
 
 The optional `networks` indicates the chains the node is interested in.
 
+The optional `compression_algorithms` is a bitfield indicating what compression
+algorithms the node supports for [gossip queries](07-routing-gossip.md#query-messages).
+
+| Bit Position  | Algorithm                                            |
+| ------------- | ---------------------------------------------------- |
+| 0             | [zlib_deflate](https://www.ietf.org/rfc/rfc1950.txt) |
```
Comment (Roasbeef) [inline 01-messaging.md:272]: Carry over from spec meeting:
  * start out w/ 2 bits 
  * bit `0` is uncompressed 
  * bit `1` is zlib

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: 01-messaging.md (hunk 3 above)
Comment (t-bast) [reply 01-messaging.md]: Done in https://github.com/lightningnetwork/lightning-rfc/pull/825/commits/5071b933d97fb7eb53f5505874c4f02d78e14028

File: 09-features.md [hunk 4]
```diff
@@ -19,40 +19,43 @@ for use of the channel, so the presentation of those features depends
...
+| 16/17 | `basic_mpp`                       | Node can receive basic multi-part payments               | IN9      | `payment_secret`          | [BOLT #4][bolt04-mpp]                                                |
+| 18/19 | `option_support_large_channel`    | Can create large channels                                | IN       |                           | [BOLT #2](02-peer-protocol.md#the-open_channel-message)              |
+| 20/21 | `option_anchor_outputs`           | Anchor outputs                                           | IN       | `option_static_remotekey` | [BOLT #3](03-transactions.md)                                        |
+| 22/23 | `option_anchors_zero_fee_htlc_tx` | Anchor commitment type with zero fee HTLC transactions   | IN       |                           | [BOLT #3][bolt03-htlc-tx], [lightning-dev][ml-sighash-single-harmful]|
+| 24/25 | `option_compression`              | Compression algorithms advertised in `init`              | IN       |                           | [BOLT #1](01-messaging.md#the-init-message)                          |
```
Comment (ysangkok) [inline 09-features.md:44]: Electrum version 4.1.0 uses bit 25 for Trampoline: https://github.com/spesmilo/electrum/blob/92226b077ad9d248e0f3325f3de89b34b1778ea5/RELEASE-NOTES#L76

Comment (t-bast) [reply 09-features.md:44]: That sounds...risky. There's no guarantee this will be the final spec feature bit.

**Comment by TheBlueMatt** [comment]
I have an implementation of this at https://github.com/rust-bitcoin/rust-lightning/pull/1015, happy to test with any other implementations :).

**Comment by t-bast** [comment]
> I have an implementation of this at rust-bitcoin/rust-lightning#1015, happy to test with any other implementations :).

I have a very old branch with early eclair support for it, but it needs many changes now, I'll get back to it when I have some time available.

**Comment by TheBlueMatt** [comment]
Needs rebase.

File: 01-messaging.md [hunk 5]
```diff
@@ -284,6 +296,9 @@ The receiving node:
     - MAY fail the connection.
   - if the feature vector does not set all known, transitive dependencies:
     - MUST fail the connection.
+  - if it has no `compression_algorithms` in common with the sending node:
+    - MUST NOT exchange messages that require compression (e.g. gossip queries).
+    - MAY fail the connection.
```
Comment (vincenzopalazzo) [inline 01-messaging.md:301]: Unity any implementation supports the compression, and these are some things that can be useful for the future. Hat do you think to change this in

```suggestion
    - MUST not fail the connection.
```

Any only after when some compression algorithms are supported or required by the communication size add "MAY"?

File: 07-routing-gossip.md [hunk 6]
```diff
@@ -636,6 +639,7 @@ The sender:
   - MUST set `chain_hash` to the 32-byte hash that uniquely identifies the chain
   that the `short_channel_id`s refer to.
   - MUST set the first byte of `encoded_short_ids` to the encoding type.
+  - MUST NOT use an encoding not supported by the remote peer.
```
Comment (vincenzopalazzo) [inline 07-routing-gossip.md:642]: also for this reason.

**Review by vincenzopalazzo — COMMENTED** [review]
LGTM, I have only some small comments

File: 01-messaging.md (hunk 5 above)
Comment (t-bast) [reply 01-messaging.md:307]: I'm sorry, I don't understand your comment at all. Can you rephrase? I think your proposed change is invalid.

Comment (vincenzopalazzo) [reply 01-messaging.md:307]: I'm sorry @t-bast, I think that there was some typo (fixed now) that makes it difficult to read the comment on the PR, and maybe a not useful question at all.

>Can you rephrase? 

Sure, I'm sorry. With my question I was pointing out I was missing the reason for failing the connection here between nodes, in other words, I was missing the case where the connection can fail here. However, I think that the case what a node may fail a connection is when the feature bit is > 0, but there is no match between compression algorithms.

I read now with another viewpoint, and I think this is the reason when a node can fail the connection?

But anyway I misunderstood some things when I read the spec. I'm so sorry.

Comment (t-bast) [reply 01-messaging.md:307]: Yes, that seems to be it. Great, I'll resolve this comment then!

**Comment by TheBlueMatt** [comment]
Needs rebase. @t-bast and I successfully tested this between LDK and eclair.

**Comment by TheBlueMatt** [comment]
Meeting discussion pivoted to just dropping zlib - it only adds so much and the complexity overhead of this PR maybe isn't worth it vs just deprecating zlib to begin with.

**Comment by t-bast** [comment]
I'll reach out to a few mobile wallets to see if they are fine with deprecating zlib.

**Comment by t-bast** [comment]
Replaced by #981 (a.k.a death to zlib)
